### PR TITLE
Force consistent MT post-redux normalisation, disallow infinite match types

### LINKED ISSUE #16596: Recursive match type fails after 3rd call when using explicit type
## Compiler version

Versions 3.2.0, 3.2.1-RC2 and 3.2.1

## Minimized code

Code below can be found [in scastie](https://scastie.scala-lang.org/qUboXwtoQceQ4VAPoCnYcw) 

```Scala
import scala.compiletime.ops.int

type Count0[N,T] <: Tuple = (N,T) match
  case (0,_)      => EmptyTuple
  case (N,String)      => String *: Count0[int.-[N, 1], String]
  case (N,Int)         => Int *: Count0[int.-[N, 1], Int]
  case (N,Float)       => Float *: Count0[int.-[N, 1], Float]  
  case (N,Double)      => Double *: Count0[int.-[N, 1], Double]  


type Count1[N,T] <: Tuple = (N,T) match
  case (0,T)      => EmptyTuple
  case (N,String)      => String *: Count1[int.-[N, 1], String]
  case (N,Int)         => Int *: Count1[int.-[N, 1], Int]
  case (N,Float)       => Float *: Count1[int.-[N, 1], Float]  
  case (N,Double)      => Double *: Count1[int.-[N, 1], Double]  

summon[Count0[1, Int] =:= Int *: EmptyTuple ]
summon[Count0[2, Int] =:= Int *: Int *: EmptyTuple] 
summon[Count0[3, Int] =:= Int *: Int *: Int *: EmptyTuple] 
summon[Count0[4, Int] =:= Int *: Int *: Int *: Int *: EmptyTuple] 
summon[Count0[5, Int] =:= Int *: Int *: Int *: Int *: Int *: EmptyTuple] 

summon[Count1[1, Int] =:= Int *: EmptyTuple ]
summon[Count1[2, Int] =:= Int *: Int *: EmptyTuple] 
// Fail from here
summon[Count1[3, Int] =:= Int *: Int *: Int *: EmptyTuple] 
summon[Count1[4, Int] =:= Int *: Int *: Int *: Int *: EmptyTuple] 
summon[Count1[5, Int] =:= Int *: Int *: Int *: Int *: Int *: EmptyTuple] 
```

## Output

```scala
[error] -- Error: /home/hmf/VSCodeProjects/sploty/meta/src/data/Data8.scala:380:62 -----
[error] 380 |    summon[Count1[3, Int] =:= Int *: Int *: Int *: EmptyTuple] 
[error]     |                                                              ^
[error]     |Cannot prove that Int *: Int *: data.Macros5.Count1[(2 : Int) - (1 : Int), Int] =:= (Int, Int, Int).
[error] -- Error: /home/hmf/VSCodeProjects/sploty/meta/src/data/Data8.scala:381:69 -----
[error] 381 |    summon[Count1[4, Int] =:= Int *: Int *: Int *: Int *: EmptyTuple] 
[error]     |                                                                     ^
[error]     |Cannot prove that Int *: Int *: data.Macros5.Count1[(3 : Int) - (1 : Int), Int] =:= (Int, Int, Int, Int).
[error] -- Error: /home/hmf/VSCodeProjects/sploty/meta/src/data/Data8.scala:382:76 -----
[error] 382 |    summon[Count1[5, Int] =:= Int *: Int *: Int *: Int *: Int *: EmptyTuple] 
[error]     |                                                                            ^
[error]     |Cannot prove that Int *: Int *: data.Macros5.Count1[(4 : Int) - (1 : Int), Int] =:= (Int, Int, Int, Int, Int).
```

## Expectation

I expects both  `Count0` and `Count1` to compile successfully. Removing the type `T` from `Count0`, to get `Count1`, seems to allow compilation to succeed, but I don't think it should change anything. Also, compilation only fails for 3 or or more recursive calls, which seems suspect.

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #16596

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -3229,7 +3229,7 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
             MatchTypeTrace.noInstance(scrut, cas, fails)
             NoType
           case MatchResult.Reduced(tp) =>
-            tp
+            tp.simplified
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: No, we should never use `simplified` in TypeComparer. `tryNormalize` is OK, simplified is not

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: There's simplified everywhere in match type reduction: in `paramInstances` and in one of the two types of reduction.  All I'm doing is making the two reduction cases the same.  That way `case (0,T) => EmptyTuple` isn't treated differently from `case (0,_) => EmptyTuple`.

Btw, in case you missed it, we addressed an issue with the use of simplified in match type reduction in #17937 by making it run under Mode.Type, to avoid how it interacts with erasure.

This one change makes reduction return a simpler type, which solves a bunch of compilation issues in match type.  It also solves a bunch of problems in test-pickler - I finally didn't have to add these test cases to the exclude list, and I would expect some more can be removed.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: Agreed that the two should behave the same. But why do we need `simplfy` (as opposed to `tryNormalize`? It was not made for this use case. The idea of `simplify` was to simplify inferred types so that they look simpler and that type inference is helped.

`simplify` can force type variables, or map type parameters to type variables, or re-arrange unions and intersections. None of this should happen at the very low level where we reduce match types. 

In short `simplify` is a method that makes sense for type _inference_, not for basic type computations.

**Comment by johnynek** [comment]
Thank you for working on this.

Personally, I think trading unbounded recursion in match types for match types composing with implicit/given resolution is a trade I would always make.

If you really want an infinite recursion in the types, you could still do so with a class or trait.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: Normalize doesn't traverse all applied types, like `(3 - 1, Int)`.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: @odersky shall I create a less powerful simplify and use that within match type reduction, or should I make all applied types try to normalize their components in tryNormalize?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: Why do we have to do it deeply at all? Can't we normalize when we inspect the element?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3232]: The `summon[Nat.FromInt[3]]` is one example which fails by not just simplifying the reductions.  But there are other examples where normalization isn't called: in Type#atoms, in TypeComparer#provablyDisjoint.

**Comment by dwijnand** [comment]
Ugh, now this is failing against the perspective community project, needing the fix in #18043...

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -3482,9 +3483,11 @@ object Types {
     private var myWidened: Type = _
 
     private def computeAtoms()(using Context): Atoms =
-      if tp1.hasClassSymbol(defn.NothingClass) then tp2.atoms
-      else if tp2.hasClassSymbol(defn.NothingClass) then tp1.atoms
-      else tp1.atoms | tp2.atoms
+      val tp1n = tp1.normalized
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:3486]: This part looks like a definite improvement; it could go in independently.

File: compiler/src/dotty/tools/dotc/typer/Inferencing.scala [hunk 3]
```diff
@@ -198,7 +198,7 @@ object Inferencing {
         case tp => foldOver(x, tp)
       }
       catch case ex: Throwable =>
-        handleRecursive("check fully defined", tp.show, ex)
+        handleRecursive("check fully defined", tp.showSummary(20), ex)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Inferencing.scala:201]: Agreed to this change.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Types.scala:3486]: @mbovel this is the other atoms fix.

I'm not sure actually if it's a `hasClassSymbol` bug.  Should `hasClassSymbol` normalize?  If I remember correctly the question, which comes from tests/pos/mt-redux-norm.perspective.scala, is does `Tuple.Fold[EmptyTuple, Nothing, [x, y] =>> x | y]` `hasClassSymbol(defn.NothingClass)`?

**Review by odersky — APPROVED** [review]
After discussions I now see that the rejections infinite match types are fine. 

I propose to change `normalized` back to `simplified` and merge the PR like that.

**Comment by soronpo** [comment]
This PR also caused the regression: https://github.com/lampepfl/dotty/issues/18261
